File: pp/pp06/QueLL/QueType.py
class NodeType:
    """ Node Type """
    def __init__(self, item):
        self.info = item
        self.next = None

class QueType:
    def __init__(self):
        self.front = None
        self.rear = None

    def make_empty(self):
        while self.front != None:
            tempPtr = self.front
            self.front = self.front.next
            del tempPtr
        self.rear = None

    def is_full(self):
        try:
            location = NodeType("test")
            del location
            return False
        except:
            return True

    def is_empty(self):
        return self.front == None

    def enqueue(self, item):
        newNode = NodeType(item)
        if self.rear is None:
            self.front = newNode
            self.rear = newNode
        else:
            temp = self.rear
            self.rear = newNode
            temp.next = newNode

    def dequeue(self):
        item = None
        if self.is_empty():
            return None
        if self.front is self.rear:
            self.rear = None
            item = self.front.info
            del self.front
            self.front = None
        else:
            node = self.front
            self.front = node.next
            item = node.info
            del node
        return item

        

File: lap/week5/04/QueType.cpp
#include "QueType.h"

QueType::QueType(int max)
// Parameterized class constructor
// Post: maxQue, front, and rear have been initialized.
//       The array to hold the queue elements has been dynamically
//       allocated.
{
  maxQue = max + 1;
  front = maxQue - 1;
  rear = maxQue - 1;
  items = new ItemType[maxQue];
}
QueType::QueType()          // Default class constructor
// Post: maxQue, front, and rear have been initialized.
//       The array to hold the queue elements has been dynamically
//       allocated.
{
  maxQue = 501;
  front = maxQue - 1;
  rear = maxQue - 1;
  items = new ItemType[maxQue];
}
QueType::~QueType()         // Class destructor
{
}

void QueType::MakeEmpty()
// Post: front and rear have been reset to the empty state.
{
  front = maxQue - 1;
  rear = maxQue - 1;
}

bool QueType::IsEmpty() const
// Returns true if the queue is empty; false otherwise.
{
  return (rear == front);
}

bool QueType::IsFull() const
// Returns true if the queue is full; false otherwise.
{
  return ((rear + 1) % maxQue == front);
}

void QueType::Enqueue(ItemType newItem)
// Post: If (queue is not full) newItem is at the rear of the queue;
//       otherwise a FullQueue exception is thrown.  
{
  if (IsFull())
    throw FullQueue();
  else
  {
    rear = (rear +1) % maxQue;
    items[rear] = newItem;
  }
}

void QueType::Dequeue(ItemType& item)
// Post: If (queue is not empty) the front of the queue has been 
//       removed and a copy returned in item; 
//       othersiwe a EmptyQueue exception has been thrown.
{
  if (IsEmpty())
    throw EmptyQueue();
  else
  {
    front = (front + 1) % maxQue;
    item = items[front];
  }
}

/*
EX 04
*/

int QueType::Length() const
{
  if (this->front < this->rear)
    return (this->rear - this->front);
  else
    return (this->maxQue - this->front + this->rear);  
}

int Length(QueType &queue)
{
  ItemType  item;
  QueType   tmp;
  int       len;

  len = 0;
  while (!queue.IsEmpty())
  {
    queue.Dequeue(item);
    tmp.Enqueue(item);
    ++len;
  }
  while (!tmp.IsEmpty())
  {
    tmp.Dequeue(item);
    queue.Enqueue(item);
  }
  return (len);
}


#include <random>
#include <ctime>
#include <iostream>
#include <string>
using namespace std;
  
int main(void)
{
  QueType   myQ;
  QueType   myQ2;
  ItemType  item;

  srand(time(NULL));

  
  
  for (int i = 0; i < 500; ++i)
  {
    item = rand() % 10;
    myQ.Enqueue(item);
  }
  for (int i = 0; i < 200; ++i)
  {
    myQ.Dequeue(item);
  }
  for (int i = 0; i < 100; ++i)
  {
    item = rand() % 10;
    myQ.Enqueue(item);
  }
  cout << "\n\n===TEST 1 : member function===" << endl;
  cout << myQ.Length() <<endl;
  cout << Length(myQ) << endl;
  return (0);
}